src/Core/Checkout/Cart/RuleLoader.php

```diff
@@ -12,12 +12,12 @@
 use Shopware\Core\Framework\Log\Package;
 use Shopware\Core\Framework\Plugin\Exception\DecorationPatternException;
 
-/**
- * @internal
```

Comment:
IMHO this should stay internal, the abstract class is not internal anymore (with this PR) nobody should use and depend on this class directly, so we can evolve it forward and keep backward compatibility on the abstract class level

Reply:
:+1: i will add @internal to the implementation, and the abstract has no @internal anymore to signalize that it can be "consumed"